src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,362 @@
+				{
+					retry[i] = true;
+					continue; // In the first phase, we try to top up existing item stacks
+				}
+
+				remaining = itemHandler.insertItem( i, remaining, false );
+				if( remaining == null )
+				{
+					break; // Awesome, full stack consumed
+				}
+			}
+
+			// If we reached this point, we still have items to insert, our first pass failed.
+			// Now we try to insert into empty slots
+			for( int i = 0; i < slotCount; i++ )
+			{
+				if( retry[i] )
+				{
+					remaining = itemHandler.insertItem( i, remaining, false );
+					if( remaining == null )
+					{
+						break; // Awesome, full stack consumed
+					}
+				}
+			}
+		}
+
+		// At this point, we still have some items left...
+		if( remaining == orgInput )
+		{
+			// The stack remained unmodified, target inventory is full
+			return iox;
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			this.onTick();
+		}
+
+		return AEItemStack.create( remaining );
+	}
+
+	@Override
+	public IAEItemStack extractItems( IAEItemStack request, Actionable mode, BaseActionSource src )
+	{
+
+		ItemStack req = request.getItemStack();
+		int remainingSize = req.stackSize;
+
+		// Use this to gather the requested items
+		ItemStack gathered = null;
+
+		final boolean simulate = ( mode == Actionable.SIMULATE );
+
+		for( int i = 0; i < itemHandler.getSlots(); i++ )
+		{
+			ItemStack sub = itemHandler.getStackInSlot( i );
+
+			if( !Platform.isSameItem( sub, req ) )
+			{
+				continue;
+			}
+
+			ItemStack extracted;
+
+			// We have to loop here because according to the docs, the handler shouldn't return a stack with size > maxSize, even if we
+			// request more. So even if it returns a valid stack, it might have more stuff.
+			do
+			{
+				extracted = itemHandler.extractItem( i, remainingSize, simulate );
+				if( extracted != null )
+				{
+					if( extracted.stackSize > remainingSize )
+					{
+						// Something broke. It should never return more than we requested... We're going to silently eat the remainder
+						AELog.warn( "Mod that provided item handler {} is broken. Returned {} items, even though we requested {}.",
+								itemHandler.getClass().getSimpleName(), extracted.stackSize, remainingSize );
+						extracted.stackSize = remainingSize;
+					}
+
+					// We're just gonna use the first stack we get our hands on as the template for the rest
+					if( gathered == null )
+					{
+						gathered = extracted;
+					}
+					else
+					{
+						gathered.stackSize += extracted.stackSize;
+					}
+					remainingSize -= gathered.stackSize;
+				}
+			}
+			while( extracted != null && remainingSize > 0 );
+
+			// Done?
+			if( remainingSize <= 0 )
+			{
+				break;
+			}
+		}
+
+		if( gathered != null )
+		{
+			if( mode == Actionable.MODULATE )
+			{
+				this.onTick();
+			}
+
+			return AEItemStack.create( gathered );
+		}
+
+		return null;
+	}
+
+	@Override
+	public TickRateModulation onTick()
+	{
+		LinkedList<IAEItemStack> changes = new LinkedList<IAEItemStack>();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Replace the type specification in this constructor call with the diamond operator ("<>"). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2293)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'changes' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)